### ISSUE TITLE
Make nn.Module cooperate with TorchDynamo

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#93735]
One idea that I have discussed with a few people and wanted to document here is having the core of PyTorch cooperate with TorchDynamo to reduce the need for guards.

Many of the guards TorchDynamo generates are checking attributes of a tree of `nn.Module()` objects.

Currently, there are two modes, controlled by the `guard_nn_modules` flag:
https://github.com/facebookresearch/torchdynamo/blob/0b1e34d53f53937b3066e61a14d210365a24b156/torchdynamo/config.py#L36
With `guard_nn_modules=True`, dynamo will check the entire nn.Module() tree. With `guard_nn_modules=False`, dynamo will just check the root `id()` of the module and assume the module structure is fixed per-instance.  `guard_nn_modules=False` matches the level of checking TorchScript does, while `guard_nn_modules=True` is a higher level of correctness guarantee.

A better way to handle this, might be to add a new `on_modified()` callback/hook to `nn.Module`.  This would be called whenever:
- eval/training mode changes
- a parameter changes
- a buffer changes
- a child module changes
- a `nn.Module().attr` changes (custom attrs, not params/buffers/modules)
- If a parameter/buffer's shape/requires_grad/etc are modified in-place (this should only check Tensor metadata, not the values)

These should all be checkable in existing places like `__setattr__`.  There are a few places that would be hard to catch, and the callback would *not* be called:
- If a python structure (like a list/dict/non-nn.Module object/etc) stored on an nn.Module is mutated.  TorchDynamo can still emit guards for these.

An open question is if this should be a global hook or a per-instance hook.  Per-instance might be more flexible, while global one would be faster since there will be one call rather than many.  (O(n) versus O(n^2) issue in number of GuardedCode()/nn.Module() instances.)

This `on_modified` hook would allow TorchDynamo to replace guards (which must be checked on the critical path on every call) with a call to [GuardedCode.invalidate()](https://github.com/facebookresearch/torchdynamo/blob/0b1e34d53f53937b3066e61a14d210365a24b156/torchdynamo/guards.py#L364), which would have zero overhead on the critical path, but add some overhead when users mutate nn.Module attributes.

TorchDynamo has something similar with the [MutationTracker](https://github.com/facebookresearch/torchdynamo/blob/0b1e34d53f53937b3066e61a14d210365a24b156/torchdynamo/mutation_guard.py#L30) class, which monkey patches user objects to get callbacks to invalidate code.  Note that currently isn't used by default.

cc @albanD @mruberry @jbschlosser @walterddr @saketh-are @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @gchanan

**Comment by albanD** [comment]
This is an interesting discussion.
At a high level I don't think this is really possible to do this for nn.Module (for the same reason we need Dynamo bytecode introspection to prevent users from hiding things from us, intentionally or not).
But I wonder what the latest development from these investigations? cc @jbschlosser ?
